[IR] Implementation of QoS Features

mm318 opened the pull request:
Implementation of QoS features in the rmw package.

mm318 committed:
initial qos interface changes

mm318 committed:
add creation of event waitable type

ross-desmond committed:
Add rmw take_event interface

mm318 committed:
add rmw event type enum

ross-desmond committed:
Modify events type from void** to rmw_event_t**

mm318 committed:
fix lint errors

dabonnie committed:
Added RMW QoS Event Definitions

dabonnie committed:
Added RMW QoS Event Definitions

dabonnie committed:
Merge branch 'rmw_events' of github.com:mm318/rmw into rmw_events

dirk-thomas committed:
add section about DCO to CONTRIBUTING.md

dabonnie committed:
Code formatting fixes

dabonnie commented on rmw/include/rmw/rmw.h:
```diff
@@ -333,7 +333,9 @@ rmw_create_subscription(
 RMW_PUBLIC
 RMW_WARN_UNUSED
 rmw_event_t *
-rmw_create_subscription_event(const rmw_subscription_t * subscription, const rmw_event_type_t event_type);
```
Add doc for this function? All of the other functions are documented.

mm318 committed:
Merge pull request #1 from mm318/rmw_events

Added RMW QoS Event Definitions

nburek commented on rmw/include/rmw/ret_types.h:
```diff
@@ -26,6 +26,14 @@ typedef int32_t rmw_ret_t;
 #define RMW_RET_ERROR 1
 #define RMW_RET_TIMEOUT 2
+/// The event is not supported by the implementation
+#define RMW_RET_EVENT_UNSUPPORTED 3
```
Bit of a nitpick, but I would just call this `RMW_RET_UNSUPPORTED`. My reasoning is that you'll return it based on the requested QoS policy not being supported as well as when checking the status of a specific event.

dabonnie replied:
Done as suggested

nburek commented on rmw/include/rmw/ret_types.h:
```diff
@@ -26,6 +26,14 @@ typedef int32_t rmw_ret_t;
+#define RMW_RET_EVENT_UNSUPPORTED 3
+#define RMW_RETCODE_PRECONDITION_NOT_MET 4
+#define RMW_RET_OUT_OF_RESOURCES 5
+#define RMW_RET_NOT_ENABLED 6
```
I didn't think ROS topics have the concept of being created in a disabled state at the moment, correct me if I'm wrong. We don't want to expose the DDS detail if it's not analogous to a ROS concept that relates to any rmw implementation.

dabonnie replied:
Ok. Check out the usage in 
```
src/ros2/rmw_connext/rmw_connext_shared_cpp/include/rmw_connext_shared_cpp/connext_static_event_info.hpp
```
 should we just return RMW_RET_UNSUPPORTED (will rename as suggested abaove) for these cases?

nburek replied:
If we're receiving back a NOT_ENABLED I think we just need to convert it to a generic `RMW_RET_ERROR` for now. I don't think it's actually a state we should get into since I think ROS always creates them enabled.

dabonnie replied:
It's a bit more involved that just a NOT_ENABLED: just attempting to read the QoS settings through RTI can generate all of the above return types. However, if we want to decouple DDS from RMW then I agree that it makes sense to remove all of the extras (4-9) and simply return RMW_RET_ERROR.

nburek commented on rmw/include/rmw/rmw.h:
```diff
@@ -223,6 +223,11 @@ rmw_create_publisher(
+RMW_PUBLIC
+RMW_WARN_UNUSED
+rmw_event_t *
+rmw_create_publisher_event(const rmw_publisher_t * publisher, const rmw_event_type_t event_type);
```
Please add documentation for all of these new functions.

nburek commented on rmw/include/rmw/rmw.h:
```diff
@@ -408,6 +420,23 @@ rmw_take_serialized_message_with_info(
   rmw_message_info_t * message_info);
 
+/*
+ * Take an event from the event handle.
```
It's tough to say without looking at the implementation, but the event_handle isn't making much sense to me yet in terms of what it does/tracks. No action on this comment yet, just place holding for me to come back.

dabonnie replied:
@nburek any further thoughts?

nburek replied:
Yeah, as I mention in other comments (maybe on other reviews). I think the naming is really confusing for some of these things.

mm318 replied:
`event_handle` has a pointer to data for the `rmw_take_event()` of the rmw implementation to use and write the correct type of information to the location pointed to by `event`. Yes, I agree the naming of the parameters could be better.

nburek commented on rmw/include/rmw/rmw.h:
```diff
@@ -417,6 +446,11 @@ rmw_create_client(
+RMW_PUBLIC
+RMW_WARN_UNUSED
+rmw_event_t *
+rmw_create_client_event(const rmw_client_t * client);
```
We aren't doing events for Services anymore, so we can remove this and the function below it.

dabonnie replied:
This is used in a few places unless my workspace is out of date.

emersonknapp replied:
We want to remove all instances of rmw_create_client_event and rmw_create_service_event - not all will have been removed yet

dabonnie replied:
👍

dabonnie replied:
Done

nburek commented on rmw/include/rmw/rmw.h:
```diff
@@ -325,6 +330,13 @@ rmw_create_subscription(
+RMW_PUBLIC
+RMW_WARN_UNUSED
+rmw_event_t *
+rmw_create_subscription_event(
```
It's not yet clear to me why the rmw layer needs to implement functions to create events and why it need separate functions for subscriptions/publishers. This could be because it's not yet clear how these are used without seeing the implementation, but it seems like they're ripe for just being copy/pasted between every rmw implementation.

mm318 replied:
The `rmw_event_t` being created needs access to the `void * data` of its corresponding `rmw_subscription_t`/`rmw_publisher_t`.

nburek commented on rmw/include/rmw/types.h:
```diff
@@ -36,6 +36,24 @@ extern "C"
+
+typedef enum rmw_event_type_t
+{
+  RMW_EVENT_SAMPLE_REJECTED,
```
I would say that all of these events outside of Liveliness and Deadlines are nice to haves and I want to make sure we're focused on getting the Liveliness and Deadline ones delivered first before we do additional work to support all of these. 

Will an application only be able to get these if they're directly calling the take functions, since there are no callbacks for them that I've seen?

dabonnie replied:
My only question for these types are does it make sense to even expose them to RMW? How much of DDS are we exposing? Maybe we should back out all of these other types and solely focus on liveliness and Deadline to decrease scope.

dabonnie replied:
Resolving as only exposing liveliness and deadline events.

nburek commented on rmw/include/rmw/types.h:
```diff
@@ -73,6 +91,13 @@ typedef struct RMW_PUBLIC_TYPE rmw_client_t
 
+typedef struct RMW_PUBLIC_TYPE rmw_event_t
+{
+  const char * implementation_identifier;
```
What is this `implementation_identifier`?

ross-desmond replied:
This is standard through all rmw types to indicate to the RMW layer which rmw implementation this event came from. My understanding is it is a safeguard.

nburek commented on rmw/include/rmw/types.h:
```diff
@@ -125,6 +150,15 @@ typedef struct RMW_PUBLIC_TYPE rmw_clients_t
+  /// The number of events represented by the array.
+  size_t event_count;
+  /// Pointer to an array of void * pointers of events.
+  rmw_event_t ** events;
```
Is the array only ever of size `event_count` or do we need something tracking the max size of it too?

ross-desmond replied:
It will only be that size for the life of this struct.

nburek commented on rmw/include/rmw/types.h:
```diff
@@ -192,14 +226,40 @@ enum RMW_PUBLIC_TYPE rmw_qos_durability_policy_t
+enum RMW_PUBLIC_TYPE rmw_qos_deadline_policy_t
+{
+  RMW_QOS_POLICY_DEADLINE_SYSTEM_DEFAULT,
+  RMW_QOS_POLICY_DEADLINE_TRANSIENT_LOCAL,
```
Are these all just copy/paste? They don't look like the values defined in the design document for any of these new policies.

dabonnie replied:
I don't know who added, need to look this up. If anyone else has thoughts please share....

emersonknapp replied:
This comment is based on an outdated version - the latest removed the copypasta and has the appropriate policy definitions - resolving thread

nburek commented on rmw/include/rmw/types.h:
```diff
@@ -192,14 +226,40 @@ enum RMW_PUBLIC_TYPE rmw_qos_durability_policy_t
+
   enum rmw_qos_reliability_policy_t reliability;
   enum rmw_qos_durability_policy_t durability;
+  enum rmw_qos_deadline_policy_t deadline;
```
These are also going to require a number specifying the amount of time for these settings.

dabonnie replied:
@nburek  can you please elaborate?

emersonknapp replied:
This is also based on an outdated version - deadline and lifespan is now an `rmw_time_t` (that's what nick meant by number)

Commit:
Change liveliness policy types to reflect actual settings

Signed-off-by: Emerson Knapp <eknapp@amazon.com>

emersonknapp committed:
Merge pull request #2 from aws-robotics/eknapp-qos-policy-types

Change liveliness policy types to reflect actual settings

Commit:
Comment out types of liveliness that will not be supported in this iteration of the feature

Signed-off-by: Emerson Knapp <eknapp@amazon.com>

ross-desmond commented on rmw/include/rmw/types.h:
```diff
@@ -192,14 +226,30 @@ enum RMW_PUBLIC_TYPE rmw_qos_durability_policy_t
+  struct rmw_time_t deadline;
+  struct rmw_time_t lifespan;
+
+  enum rmw_qos_liveliness_policy_kind_t liveliness;
```
rmw_qos_liveliness_policy_kind_t is a different convention than rmw_qos_durability_policy_t. Should be rmw_qos_liveliness_policy_t.

emersonknapp replied:
Durability is only an enum, liveliness is two things, a `kind` and a `lease_duration`, you think we should create a struct instead to carry both items?  The reason that I split it is because `history` QoS was already split between the `enum rmw_qos_history_policy_t` and `size_t depth` members - I was following the standing convention

nburek replied:
I don't think we need to create a new wrapper stuct. I think Ross' comment was mainly about the naming of the enum type.

mm318 replied:
Okay, let's rename `rmw_qos_liveliness_policy_kind_t` to `rmw_qos_liveliness_policy_t` to be consistent with `rmw_qos_history_policy_t`.

nburek commented on rmw/include/rmw/types.h:
```diff
@@ -73,6 +91,13 @@ typedef struct RMW_PUBLIC_TYPE rmw_client_t
+typedef struct RMW_PUBLIC_TYPE rmw_event_t
+{
+  const char * implementation_identifier;
+  void * data;
```
What is in this `data`? Looking at the review for rcl makes me think it's not just information about an event but all sorts of additional metadata. Is this just some generic pointer to a bunch of data used by the rmw_impl?

dabonnie replied:
One of the examples I see for data is the ConnextCustomEventInfo in rmw_connext. My guess is this has to be void * due to the different DDS implementations? It might be worth having a discussion to abstract this into a concrete type rather than void *.

nburek replied:
So is it a pointer to reference data for the rmw implementation to use or is it a pointer to data for things above the rmw layer to use? No documentation, no data type, and a pretty generic name make it hard to understand.

dabonnie replied:
> So is it a pointer to reference data for the rmw implementation to use or is it a pointer to data for things above the rmw layer to use? No documentation, no data type, and a pretty generic name make it hard to understand.

I don't disagree, as the above seems to be the theme for this feature. Right now I'm confusing this with the void * event found in rmw_connext (I don't think we need void * there), so need to figure out what should be done here.

mm318 replied:
It is a pointer to data for the rmw implementation to use, not for things above the `rmw` layer to use.

ross-desmond commented on rmw/include/rmw/types.h:
```diff
@@ -192,14 +226,30 @@ enum RMW_PUBLIC_TYPE rmw_qos_durability_policy_t
+  struct rmw_time_t lifespan;
+
+  enum rmw_qos_liveliness_policy_kind_t liveliness;
+  struct rmw_time_t liveliness_lease_duration;
```
Could we add this to the rmw_qos_liveliness_policy_t since they are both required and related?

emersonknapp replied:
See above comment about the existing pattern of splitting HISTORY policy - if we want to group it's fine with me, but just want you to be aware of that while evaluating this

mm318 replied:
Okay, let's be consistent with `rmw_qos_history_policy_t` + `size_t`, and _not_ group `rmw_qos_liveliness_policy_kind_t` + `rmw_time_t` into `rmw_qos_liveliness_policy_t`.

ross-desmond commented on rmw/include/rmw/qos_profiles.h:
```diff
@@ -22,12 +22,20 @@ extern "C"
 
 #include "rmw/types.h"
 
+#define rmw_qos_deadline_default {0, 0}
```
#define default should follow convention, all caps.

ross-desmond commented on rmw/include/rmw/types.h:
```diff
@@ -239,6 +289,154 @@ typedef enum RMW_PUBLIC_TYPE
 } rmw_log_severity_t;
 
+/// Defined in the DDS SSpec 15-04-10 section 2.2.4.1 Communication Status
+typedef struct RMW_SAMPLE_LOST_STATUS
```
typedef struct <CAPS> does not follow conventions
https://google.github.io/styleguide/cppguide.html#Type_Names

ross-desmond reviewed (changes requested):
Overall fairly good, please run ament_uncrustify and ament_ccplint before posting a review :)

Commit:
Re-enable manual liveliness, name to by_node

dabonnie committed:
Fixed style issues
Removed unsupported QoS types

dabonnie commented:
> Overall fairly good, please run ament_uncrustify and ament_ccplint before posting a review :)

Interestingly enough many of your style comments weren't caught by the linters. Appreciate the catches and the style guide links.

dabonnie commented on rmw/include/rmw/types.h:
```diff
@@ -36,6 +36,18 @@ extern "C"
 #define RMW_GID_STORAGE_SIZE 24
 
+/// Define QoS policy events
+typedef enum rmw_event_type_t
```
This is incomplete. Really RMW_EVENT_SUBSCRIPTION_MATCHED and RMW_EVENT_OFFERED_INCOMPATIBLE_QOS should be removed as well. @ross-desmond this affects your map lookup in rmw_connext though. Seem reasonable for this removal?

dabonnie committed:
Removed extra QoS types

mm318 committed:
add an invalid enum in rmw_event_type_t for variables to default to

mm318 commented:
OSRF requested that for new `rmw` types, we should be using the `init()`/`fini()` pattern (similar to what is found in `rcl`), instead of the `create()`/`destroy()` pattern (https://github.com/ros2/rmw/pull/167#discussion_r269694583). I will make this change.

mm318 committed:
change rmw_event_t APIs from create()/destroy() to init()/fini() pattern

ross-desmond committed:
Move rmw_*_event_init implementation to rmw

**Summary**
Event initialization should require no modifications from the underlying rmw layer and therefore can be initialized in the rmw layer.

ross-desmond commented:
I've added the interface changes for rmw_*_event_init

ross-desmond committed:
Fix uncrustify divergence in event.c

ross-desmond committed:
Remove erroneous rmw_delete_event

mm318 committed:
Merge remote-tracking branch 'origin/master' into qos
